File: 프로그래머스/2/42587. 프로세스/프로세스.swift

// 문제는 location에 위치한 값이 dequeue 되는 순간을 어떻게 측정할 것인가 ?
// 쉽다 cnt 변수를 두면 된다
// dequeue 할 때마다 cnt를 하나씩 올린다
// location 위치만 true이고, priorities와 길이가 같은 Bool 배열을 만들어서, priorities 배열과 똑같이 queue를 돌리고, dequeue 될 때 해당 위치가 true인 경우에 cnt를 출력하도록 한다.

// priorities를 순차적으로 탐색하며, dequeue한 값이 max 값과 일치하면 버린다. cnt += 1
// 일치하지 않으면(더 작으면) 다시 enqueue 한다
// 이것을 Bool 배열에도 똑같이 적용한다
// true인 index가 버려지면 cnt를 return 한다


import Foundation

func solution(_ priorities:[Int], _ location:Int) -> Int {
    
    var priorities = priorities
    var location = location
    var result = 0
    
    while priorities.count > 0 {
        location -= 1
        let maxVal = priorities.max()!
        let front = priorities[0]
        if front != maxVal {
            priorities.append(front)
            priorities.removeFirst()
            if location < 0 {
                location = priorities.count - 1
            }
        } else {
            result += 1
            priorities.removeFirst()
            if location < 0 { break }
        }
    }
    
    return result
}

File: 프로그래머스/2/42587. 프로세스/프로세스.py
# 문제 분석
# priorities 배열은 우선순위(원소의 값은 큰 순서)대로 제거해야 한다.
# 일종의 원형 큐처럼 동작한다.
# 현재 배열에서 가장 우선 순위가 높은 원소를 만날 때까지, 앞에서 제거 및 뒤에 추가를 반복한다.
# location 위치의 원소가 실행되는 순서를 구하라. (위치는 0부터, 순서는 1부터)

# 접근
# priorities 배열을 탐색하면서, [priority, index] 형태의 priorityWithIndex 배열을 생성한다.
# priorities 배열을 오름차순 정렬한 sortedPriorities 배열을 생성한다.
# while priorityWithIndex:
    # top = priorityWithIndex.popleft()
    # if top[0] == sortedPriorities[-1]: # 정상 실행
        # cnt += 1
        # sortedPriorities.pop()
        # if top[1] == location:
            # return cnt
    # else: # 다시 큐에 삽입
        # priorityWithIndex.append(top)

from collections import deque
    
def solution(priorities, location):
    
    cnt = 0
    
    priorityWithIndex = deque()
    for i in range(len(priorities)):
        priorityWithIndex.append([priorities[i], i])
        
    sortedPriorities = sorted(priorities)
    
    while priorityWithIndex:
        top = priorityWithIndex.popleft() # top = [priority, index]
        if top[0] == sortedPriorities[-1]: # 정상 실행
            cnt += 1
            sortedPriorities.pop()
            if top[1] == location:
                return cnt
        else: # 다시 큐에 삽입
            priorityWithIndex.append(top)